### PR TITLE
Allow to "use" minor versions

### LINKED ISSUE #360: func-e use latest patch
## Current State
Right now, we do a lazy lookup of the most recent version of envoy, as determined by https://archive.tetratelabs.io/envoy/envoy-versions.json latestVersion Whatever the latest version is at the time is serialized to ~/.func-e/version to prevent excessive downloading and ensure the config is stable.

## Impact
This is great because the first time use of Envoy doesn't require knowing any version, much less the patch version. However, they also won't automatically get a new patch version, and usually patches are desirable. Moreover, there's no way to get the latest patch of an older minor version. For example, if 1.20 is the last minor, there's no way for a user to automatically get the last patch of 1.18.

## Current Workaround

Currently, if a user wants to get the latest patch, they can delete ~/.func-e/version and have it re-fetch. To get the latest patch of an older minor (ex 1.18), they need to parse `func-e versions -a` and `func-e use` that value (ex. 1.18.4). While this can be worked around manually, it is even more a pain in devops scenarios like build matrixes.

## Design idea

It is a better default to get the latest patch of the latest minor. For example, at the cost of re-hitting envoy-versions.json each time, we can take a value such as '1.19' and translate it into the concrete latest version, even after '1.20' comes out. Basically, if we see an expression missing the patch part, do a pre-emptive fetch of the version list and parse the last patch from that. After that point, the existing logic kicks in to see if it is already downloaded or not.

In the first implementation, it isn’t necessary for us to cache the envoy-versions.json to disk even if we do in process, but we should handle the offline case and fall-back to an existing patch when the versions list is unavailable. There should be a test for this edge case.

### PR DESCRIPTION
**Pull request by dio**
This changeset introduces a way to switch the Envoy version using compatible
Envoy version (for example 1.18) and automatically infer the available
patched version to be downloaded. For instance, if the currently
downloaded version is 1.18.3, and the remote one has 1.18.4, calling
"use" downloads and sets the current version to 1.18.4.

Fixes #360 

Signed-off-by: Dhi Aurrahman <dio@tetrate.io>

### CODE REVIEW

**Comment by dio** [comment]
Still missing the test for the edge case, when `FuncEVersions` failed.

**Comment by dio** [comment]
Kapture 2021-08-30 at 21 44 57

File: internal/version/last_known_compatible_envoy.txt
Comment (codefromthecrypt) [inline internal/version/last_known_compatible_envoy.txt:1]: let's derive this from last_known so there's less maintenance. Remember, the last_known is only used in docs. It should never be thought of as the actual latestVersion. So, what I mean is if this is for purpose of docs, stripping the minor from last_known_envoy should do the trick!

Comment (dio) [reply internal/version/last_known_compatible_envoy.txt:1]: Yes, sure thing!

**Comment by dio** [comment]
Oops. I have wrongly sent out the wrong GIF haha. Will update soon.

File: internal/envoy/install.go
Comment (codefromthecrypt) [inline internal/envoy/install.go:84]: mention which subsequent steps need this? I hope we don't have any and you can delete this. The main reason for the version is just to get the install path, which this method returns. If you delete this, what breaks?

File: internal/envoy/install.go
Comment (codefromthecrypt) [inline internal/envoy/install.go:100]: both the logic here and CompatibleVersionStrings and its corresponding tests feel like too much code for the job. Also it is looser than it needs to be. There's only one call site as far as I can tell (here), so if you define logic here until there are multiple call sites. That way the function can be un exported and focus only on getting the thing.. iterative seems simplest. I think this can reduce the lines and things to explain, test, review during future refactorings as FuncEVersions can be left alone.

```
// findLatestVersion returns the latest available patch or empty if not found
findLatestVersion(ev version.ReleaseVersions, minor string) string {
var latestPatch
for _, patch ... {
  if patch has prefix minor && patch > latestPatch {
latestPath = patch
}
}
return latestPatch
```

**Review by codefromthecrypt — COMMENTED** [review]
Thanks for the start on this! and you are fast! Appreciate the preview and I understand this was marked draft so not in final form. I went ahead and made some notes anyway :D

So, main thing to focus is on least impact to achieve the task. I made some suggestions to tighten up the change. Each function defined has to pull its weight once merged.

**Comment by codefromthecrypt** [comment]
ps cool screen recording.. what do you use?

**Comment by dio** [comment]
> ps cool screen recording.. what do you use?

@codefromthecrypt I use https://getkap.co/. I know cmd+shift+5 on macOS gives you screen recording and [kap](https://getkap.co/) is an electron-based app. I'm too lazy to convert the recorded .mov to GIF and put it inside the clipboard: [kap](https://getkap.co/) has this functionality with a single click.

image

File: e2e/func-e_use_test.go
Comment (dio) [reply e2e/func-e_use_test.go:30]: Added.

File: internal/envoy/install.go
Comment (dio) [inline internal/envoy/install.go:93]: @codefromthecrypt do we have a facility to inject FuncEVersions in a test?

**Comment by dio** [comment]
re: Failing downloading versions after install. Seems like there is no way to overcome this? 

So the scenario of fallback to the "current version" is _only_ when the expected binary version is available at least one? For example: 

```
$ cat ~/.func-e/version
1.18.3
$ ls ~/.func-e/versions/1.18.3/
bin/envoy
$ # network failure happens after this line
$ func-e use "1.18"
1.18.3 is already downloaded
```

But what if:

```
$ cat ~/.func-e/version
1.18.3
$ ls ~/.func-e/versions/1.18.3/
bin/envoy
$ # network failure happens after this line
$ func-e use 1.19
```

Are we going to "use" 1.18.3 or throw an error?

**Comment by codefromthecrypt** [comment]
on using a different minor when network isn't available, that should be a legit error

main thing is we are trying to allow offline when the minor is the same

File: internal/envoy/install.go
Comment (codefromthecrypt) [reply internal/envoy/install.go:93]: To answer your question directly, the cmd package already has a mock server (in `setupTest()`), but we shouldn't need that in this package. This file should be left alone. InstallIfNeeded has no idea of the current version file or anything. Its only responsibility is installing a concrete version which it has been passed.

If new logic should be in this package, it would be version.go, and only add a new (single method not broken up into many as that makes it hard to understand). Iterating through a list of versions to get the patch is only several lines.


I think the key insight here will be that a compositional approach is simpler both to review and implement. IOTW, there's logic in places like `run.go` which would have a new compositional function call to resolve the version into a patch.

```go
		if err := setHomeEnvoyVersion(c.Context, o); err != nil {
			return err
		}
		v, _, err := envoy.CurrentVersion(o.HomeDir)
		if err != nil {
			return NewValidationError(err.Error())
		}
+		if minor pattern matches v {
+			ev, err = envoy.FuncEVersions(ctx, o.EnvoyVersionsURL, o.Platform, o.Version)
+			if err != nil {
+				return NewValidationError(err.Error())
+			}
+			latest = findLatestVersion(ev, v)
+			if latest == "" {
+				return NewValidationError("no version available")
+			}
+			o.EnvoyVersion = latest
+                 } else {
+			o.EnvoyVersion = v
+                 }
-		o.EnvoyVersion = v
```
Bear in mind the above is a sketch and this whole piece may be better off as a higher order function.

If it ends up needing to do be extracted into a higher order function, I personally would slightly adjust FuncEVersions so that it is an interface. That way you don't need to deal with re-testing it in cmd at least the http related stuff.

Ex FuncEVersions is then an interface with 2 functions, the existing as `Get() (version.ReleaseVersions, error)` and a new one to support this as `FindLatestVersion(v) (string, error)`

The struct implementing it can optionally cache between these calls but we should do that separately to keep review burden down. Restructuring to what I mentioned I would expect only 20 lines or so change excluding indents.

The diff changes slightly as you then have one less visible if statement at the call site
```go
		if err := setHomeEnvoyVersion(c.Context, o); err != nil {
			return err
		}
		v, _, err := envoy.CurrentVersion(o.HomeDir)
		if err != nil {
			return NewValidationError(err.Error())
		}
+		if minor pattern matches v {
+			// since this is lazy it probably will be initialized sooner and ideally plumbed through, but again that can be a separate change.
+			feV := envoy.NewFuncEVersions(ctx, o.EnvoyVersionsURL, o.Platform, o.Version)
+			latest, err = feV.FindLatestVersion(v)
+			if latest == "" {
+				return NewValidationError(err.Error())
+			}
+			o.EnvoyVersion = latest
+                 } else {
+			o.EnvoyVersion = v
+                 }
-		o.EnvoyVersion = v
```

Anyway, it pains me to ask you to change the code, but I think in any case code isn't something to hold tightly, I've deleted thousands of lines I wrote here as well tens of thousands of lines I didn't write :D Main thing is that if we focus on responsibilities, I think this change can be implemented changing less existing code and most of that will come from composition.

Hope this helps!

Comment (dio) [reply internal/envoy/install.go]: I have added an implementation of FuncEVersions to have the `FindLatestVersion` method. Please take a look. If that makes sense I'll continue to put it in cmd.

File: internal/envoy/versions.go
Comment (codefromthecrypt) [inline internal/envoy/versions.go:82]: it makes me uncomfortable a little to pass in version.Version and pass out a string. I think we shouldn't enter this at all if we already know we have a concrete version. So, if anything it should pass out version.Version and in minorVersion

Also, we tend to use pointer style
```suggestion
// FindLatestPatch finds the latest patch version for the given minor or raises an error.
// minor is just like envoy.Version format, except missing the patch. Ex. 1.18 or 1.20_debug
func (f *funcEVersions) FindLatestPatch(minorVersion string) (version.Version, error) {
```

File: internal/envoy/versions.go
Comment (codefromthecrypt) [inline internal/envoy/versions.go:83]: ex we shouldn't enter this case at all, there's no reason to assume our internal code will pass the wrong value, and if we did we'd have to test it, so it isn't really fun

This type shouldn't even need to be instantiated if someone passes a concrete version iotw.

File: internal/envoy/versions.go
Comment (codefromthecrypt) [inline internal/envoy/versions.go:135]: this isn't worth scrolling around for, can you inline the error check instead of making another function then having to name it and scroll to it?

File: internal/envoy/versions.go
Comment (codefromthecrypt) [inline internal/envoy/versions.go:128]: I would rather spend the code on this being lazy vs it being a swappable implementation, especially if there's no user that swaps it. How about just initializing it so it is never null if using New?

**Review by codefromthecrypt — COMMENTED** [review]
I think this is on the right track, just I think the new code shouldn't be called at all unless we actually need to (they specified a minor version)

ex getLatestPatch(alreadyAPatch) isn't sensible.

Otherwise, I think this is on the right track. thanks

File: internal/envoy/versions.go
Comment (dio) [reply internal/envoy/versions.go:128]: Thanks for this. Was not thinking thoroughly about this.

File: internal/envoy/versions.go
Comment (dio) [reply internal/envoy/versions.go:83]: Got it.

**Comment by mathetake** [comment]
Let me know when this is ready for review! 👯

File: internal/envoy/versions.go
Comment (codefromthecrypt) [inline internal/envoy/versions.go:39]: speaking aloud as going to be offline 🌴 one neat thing about this being an interface is that code to conditionally call FindLatestPatch is easy to guard in the cmd package.

Ex. you could have code that accepts this interface and the test cases will be easy to write as the impl can raise an error on Get() and do whatever you are testing on FindLatestVersions both without needing to worry about http calls and not requiring to bring most version parsing internal here.

**Comment by dio** [comment]
@mathetake sure. Thank you! I'm looking at the e2e installing with the minor and offline cases.

**Comment by dio** [comment]
@mathetake @codefromthecrypt I think I'll split the PR by first introducing FuncEVersions as an interface. After that, I'll update the "use" functionality. WDYT?

**Comment by mathetake** [comment]
sg!

File: internal/cmd/use_test.go
Comment (dio) [inline internal/cmd/use_test.go:155]: Please let me know if you want me to wrap some of these. I don't know the preference for wrapping codes for tests. Some projects asked me to wrap, but others want it easy to read, hence copy-pasting is fine.

**Comment by dio** [comment]
> Let me know when this is ready for review! 👯

When you have time, please take a look. Thanks!

**Comment by codefromthecrypt** [comment]
I hereby relinquish review to the person who reviewed all my work so far: @mathetake

In other words, when Tasheshi and you feel ready, go ahead and merge. ciao until early next week 🌴 🏖️

File: internal/cmd/app.go
Comment (mathetake) [inline internal/cmd/app.go:92]: I think we'd better put comment about this brunch! - this only happens in tests, right?

Comment (mathetake) [reply internal/cmd/app.go:92]: sorry typo:  *branch 😄

File: internal/cmd/run.go
Comment (mathetake) [inline internal/cmd/run.go:159]: maybe more friendly to mention the fact that "using the latest among installed ones"?

File: e2e/func-e_use_test.go
Comment (mathetake) [inline e2e/func-e_use_test.go:155]: nit: strictly speaking this is not "min", right? 😄

Comment (dio) [reply e2e/func-e_use_test.go:155]: Yeah, I'm changing this to `0` to make it more "versatile", but still set the purpose right.

File: internal/cmd/run.go
Comment (dio) [reply internal/cmd/run.go:159]: Got it.

File: internal/cmd/app.go
Comment (dio) [reply internal/cmd/app.go:93]: I got it @mathetake! I'm hungry already!

File: internal/cmd/run.go
Comment (dio) [reply internal/cmd/run.go]: Updated. See what you think.

File: e2e/func-e_use_test.go
Comment (dio) [reply e2e/func-e_use_test.go]: Updated.

File: internal/cmd/use.go
Comment (mathetake) [inline internal/cmd/use.go:100]: May we could avoid allocating the string in every loop 😄 
```suggestion
	prefix := string(minorVersion)+"."
	for i := range rows {
		if strings.HasPrefix(string(rows[i].version), prefix) {
```

Comment (mathetake) [reply internal/cmd/use.go:100]:
```suggestion
	for _, row := range rows {
		if strings.HasPrefix(string(row.version), string(minorVersion)+".") {
```

Comment (dio) [reply internal/cmd/use.go]: Ah OK

File: internal/globals/globals.go
Comment (mathetake) [inline internal/globals/globals.go:98]: I think we might want to tests to make sure this regex is actually correct to avoid future drifts?

Comment (mathetake) [reply internal/globals/globals.go:98]: I should have commented this for other regexes to @codefromthecrypt 😄

Comment (dio) [reply internal/globals/globals.go:98]: I can do that.

**Comment by mathetake** [comment]
maybe we should modify this instruction as something like `should look like "1.19.1" or "1.19"`?
```
$ ./func-e use 1.19.
invalid [version] argument: "1.19." should look like "1.19.1"
show usage with: func-e help
```

File: internal/globals/globals.go
Comment (mathetake) [reply internal/globals/globals.go:98]: thank you! 🎸

**Comment by dio** [comment]
> maybe we should modify this instruction as something like `should look like "1.19.1" or "1.19"`?
> 
> ```
> $ ./func-e use 1.19.
> invalid [version] argument: "1.19." should look like "1.19.1"
> show usage with: func-e help
> ```

Yes, I missed the USAGE.md part. Will update it.

File: internal/cmd/run.go
Comment (mathetake) [reply internal/cmd/run.go]: Brilliant!

File: internal/cmd/use_test.go
Comment (mathetake) [inline internal/cmd/use_test.go:170]:
```suggestion
	// The initial version to be installed.
	minorVersion := "1.12"
	latestPatch := "3"
	initial := avaliableVersions{
		latestPatch:  latestPatch,
		versions:    []version.Version{minorVersion},
	}
```

File: internal/cmd/use_test.go
Comment (mathetake) [reply internal/cmd/use_test.go:155]: I think this looks fine as-is! Thank you!

**Review by mathetake — APPROVED** [review]
Mostly LGTM now!

File: internal/cmd/use_test.go
Comment (dio) [reply internal/cmd/use_test.go:170]: Love this. Thanks!

File: internal/cmd/use_test.go
Comment (mathetake) [reply internal/cmd/use_test.go:155]: I think this looks fine as-is! Thank you!

File: internal/cmd/use_test.go
Comment (mathetake) [reply internal/cmd/use_test.go:170]: oh my bad - nvm 😄

**Comment by dio** [comment]
Thanks for your time, @mathetake! If you have other comments, please do so. Otherwise, I'll merge it in when the CI jobs are green 🙏🏽

File: internal/globals/globals_test.go
Comment (mathetake) [inline internal/globals/globals_test.go:91]:
```suggestion
		if matched = globals.EnvoyMinorVersionPattern.FindAllStringSubmatch(tc.version, -1); matched != nil {
```

Comment (dio) [reply internal/globals/globals_test.go:91]: Oops.

**Review by mathetake — APPROVED** [review]
Awesome!

**Comment by mathetake** [comment]
Good to go!

**Comment by dio** [comment]
Thank you @mathetake @codefromthecrypt!

**Pull request merged**
